File: assets/js/engine/validator.js
/**
 * Validación de factibilidad — puerto JS de `backend/app/solver/validator.py`.
 *
 * Corre DESPUÉS de asignar tutores y ANTES de la búsqueda. Detecta en milisegundos
 * lo que ningún backtracking puede arreglar y lo explica en el idioma del director.
 *
 * error   -> se aborta la búsqueda (status: infeasible)
 * warning -> se busca igual, pero se avisa
 *
 * @module engine/validator
 */

import { groupLabel } from './contract.js';

const conflict = (severity, code, message, extra = {}) => ({
  severity, code, message,
  group_id: null, subject_id: null, teacher_id: null, missing_hours: null,
  ...extra,
});

export function validate(ctx) {
  return [
    ...checkGroupCapacity(ctx),
    ...checkEligibility(ctx),
    ...checkMaxPerDay(ctx),
    ...checkSubjectCapacity(ctx),
    ...checkGlobalCapacity(ctx),
    ...checkTeacherAvailability(ctx),
  ];
}

export const hasBlockingErrors = (conflicts) => conflicts.some((c) => c.severity === 'error');

// --------------------------------------------------------------------------- //
function checkGroupCapacity(ctx) {
  const out = [];
  for (const gid of ctx.groupOrder) {
    const group = ctx.groups.get(gid);
    const capacity = ctx.groupSlots.get(gid).size;
    const demand = group.curriculum.reduce((acc, c) => acc + c.weekly_hours, 0);
    if (demand > capacity) {
      out.push(conflict('error', 'GROUP_OVER_CAPACITY',
        `El grupo ${groupLabel(group)} pide ${demand} h a la semana pero su rejilla sólo tiene ` +
        `${capacity} espacios disponibles. Reduce horas del plan o agrega bloques al horario.`,
        { group_id: gid, missing_hours: demand - capacity }));
    } else if (demand < capacity * 0.5) {
      out.push(conflict('info', 'GROUP_UNDERUSED',
        `El grupo ${groupLabel(group)} sólo ocupa ${demand} de ${capacity} espacios; quedarán muchas horas libres.`,
        { group_id: gid }));
    }
  }
  return out;
}

function checkEligibility(ctx) {
  const out = [];
  for (const gid of ctx.groupOrder) {
    const group = ctx.groups.get(gid);
    const gSlots = ctx.groupSlots.get(gid);
    for (const entry of group.curriculum) {
      if (!entry.weekly_hours) continue;
      const elig = ctx.eligible.get(`${gid}|${entry.subject_id}`) || [];
      const subjectName = ctx.subjects.get(entry.subject_id)?.name || entry.subject_id;

      if (!elig.length) {
        const reason = entry.assign_to_tutor
          ? 'el grupo no tiene tutor asignado'
          : 'ningún profesor la tiene en su lista de materias';
        out.push(conflict('error', 'NO_ELIGIBLE_TEACHER',
          `${subjectName} en ${groupLabel(group)}: ${reason}. Asigna la materia a algún profesor.`,
          { group_id: gid, subject_id: entry.subject_id, missing_hours: entry.weekly_hours }));
        continue;
      }

      const usable = new Set();
      for (const tid of elig) {
        const tSlots = ctx.teacherSlots.get(tid);
        for (const slot of gSlots) if (tSlots.has(slot)) usable.add(slot);
      }
      if (!usable.size) {
        out.push(conflict('error', 'NO_COMMON_SLOT',
          `${subjectName} en ${groupLabel(group)}: la disponibilidad de sus profesores no coincide ` +
          'con ningún horario hábil del grupo.',
          { group_id: gid, subject_id: entry.subject_id, missing_hours: entry.weekly_hours }));
      } else if (usable.size < entry.weekly_hours) {
        out.push(conflict('error', 'NOT_ENOUGH_COMMON_SLOTS',
          `${subjectName} en ${groupLabel(group)} necesita ${entry.weekly_hours} h pero sólo existen ` +
          `${usable.size} horarios compatibles con sus profesores.`,
          { group_id: gid, subject_id: entry.subject_id, missing_hours: entry.weekly_hours - usable.size }));
      }
    }
  }
  return out;
}

function checkMaxPerDay(ctx) {
  const out = [];
  const nDays = ctx.nDays;
  for (const gid of ctx.groupOrder) {
    const group = ctx.groups.get(gid);
    for (const entry of group.curriculum) {
      const ceiling = entry.max_per_day * nDays;
      if (entry.weekly_hours > ceiling) {
        const subjectName = ctx.subjects.get(entry.subject_id)?.name || entry.subject_id;
        out.push(conflict('error', 'MAX_PER_DAY_TOO_LOW',
          `${subjectName} en ${groupLabel(group)}: ${entry.weekly_hours} h semanales no caben con un ` +
          `tope de ${entry.max_per_day} h/día en ${nDays} días.`,
          { group_id: gid, subject_id: entry.subject_id, missing_hours: entry.weekly_hours - ceiling }));
      }
    }
  }
  return out;
}

function checkSubjectCapacity(ctx) {
  const out = [];
  const demand = new Map();
  const boundToTutor = new Map();
  for (const gid of ctx.groupOrder) {
    for (const entry of ctx.groups.get(gid).curriculum) {
      demand.set(entry.subject_id, (demand.get(entry.subject_id) || 0) + entry.weekly_hours);
      if (entry.weekly_hours) {
        const prev = boundToTutor.get(entry.subject_id);
        boundToTutor.set(entry.subject_id, prev === undefined ? entry.assign_to_tutor : prev && entry.assign_to_tutor);
      }
    }
  }

  for (const [sid, hours] of demand) {
    if (!hours || boundToTutor.get(sid)) continue;   // las tutorías se validan por elegibilidad
    let capacity = 0;
    let slotCapacity = 0;
    for (const tid of ctx.teacherOrder) {
      if (ctx.teachers.get(tid).subject_ids.includes(sid)) {
        capacity += ctx.teacherMaxWeekly.get(tid);
        slotCapacity += ctx.teacherSlots.get(tid).size;
      }
    }
    const effective = Math.min(capacity, slotCapacity);
    const subjectName = ctx.subjects.get(sid)?.name || sid;
    if (effective < hours) {
      out.push(conflict('error', 'SUBJECT_CAPACITY',
        `${subjectName}: la escuela requiere ${hours} h semanales y sus profesores sólo pueden cubrir ` +
        `${effective} h (carga máxima ${capacity} h, disponibilidad ${slotCapacity} h). ` +
        'Contrata apoyo o baja las horas del plan.',
        { subject_id: sid, missing_hours: hours - effective }));
    } else if (effective < hours * 1.1) {
      out.push(conflict('warning', 'SUBJECT_CAPACITY_TIGHT',
        `${subjectName} va muy justa: ${hours} h requeridas contra ${effective} h disponibles. ` +
        'Es probable que falten horas por acomodar.',
        { subject_id: sid }));
    }
  }
  return out;
}

function checkGlobalCapacity(ctx) {
  const demand = ctx.requiredHours;
  let capacity = 0;
  for (const cap of ctx.teacherMaxWeekly.values()) capacity += cap;

  if (demand > capacity) {
    return [conflict('error', 'GLOBAL_CAPACITY',
      `La escuela requiere ${demand} horas-clase semanales y la plantilla suma ${capacity} horas ` +
      `contratadas. Faltan ${demand - capacity} h.`,
      { missing_hours: demand - capacity })];
  }
  if (demand > capacity * 0.95) {
    return [conflict('warning', 'GLOBAL_CAPACITY_TIGHT',
      `Se usará más del 95 % de la plantilla (${demand}/${capacity} h). Con tan poco margen es normal ` +
      'que el horario quede con huecos difíciles.')];
  }
  return [];
}

function checkTeacherAvailability(ctx) {
  const out = [];
  for (const tid of ctx.teacherOrder) {
    const teacher = ctx.teachers.get(tid);
    const available = ctx.teacherSlots.get(tid).size;
    if (teacher.max_weekly_hours > 0 && available === 0) {
      out.push(conflict('warning', 'TEACHER_NO_AVAILABILITY',
        `${teacher.name} tiene ${teacher.max_weekly_hours} h contratadas pero su matriz de ` +
        'disponibilidad no deja ningún horario libre.',
        { teacher_id: tid }));
    } else if (available < teacher.max_weekly_hours) {
      out.push(conflict('info', 'TEACHER_AVAILABILITY_BELOW_CONTRACT',
        `${teacher.name} sólo está disponible ${available} h de las ${teacher.max_weekly_hours} h contratadas.`,
        { teacher_id: tid }));
    }
  }
  return out;
}

File: backend/app/solver/validator.py
"""
Validación de factibilidad (pre-vuelo).

Corre DESPUÉS de asignar tutores y ANTES de la búsqueda. Su trabajo es detectar en
milisegundos los problemas que ninguna cantidad de backtracking puede resolver, y
explicarlos en español para el director de la escuela —no para el programador.

Distingue:
  ERROR   -> imposible por construcción; se aborta la búsqueda (`status: infeasible`).
  WARNING -> muy probable que falten horas; se busca igual y se reporta al final.
"""

from __future__ import annotations

from collections import defaultdict
from typing import Dict, List, Tuple

from ..models.schemas import Conflict, Severity
from .domain import ProblemContext


def validate(ctx: ProblemContext) -> List[Conflict]:
    conflicts: List[Conflict] = []
    conflicts += _check_group_capacity(ctx)
    conflicts += _check_eligibility(ctx)
    conflicts += _check_max_per_day(ctx)
    conflicts += _check_subject_capacity(ctx)
    conflicts += _check_global_capacity(ctx)
    conflicts += _check_teacher_availability(ctx)
    return conflicts


def has_blocking_errors(conflicts: List[Conflict]) -> bool:
    return any(c.severity is Severity.ERROR for c in conflicts)


# --------------------------------------------------------------------------- #
def _check_group_capacity(ctx: ProblemContext) -> List[Conflict]:
    """El plan de estudios no puede pedir más horas que celdas tiene el grupo."""
    out: List[Conflict] = []
    for gid in ctx.group_order:
        group = ctx.groups[gid]
        capacity = len(ctx.group_slots[gid])
        demand = group.total_weekly_hours
        if demand > capacity:
            out.append(
                Conflict(
                    severity=Severity.ERROR,
                    code="GROUP_OVER_CAPACITY",
                    message=(
                        f"El grupo {group.label} pide {demand} h a la semana pero su rejilla "
                        f"sólo tiene {capacity} espacios disponibles. Reduce horas del plan o "
                        "agrega bloques al horario."
                    ),
                    group_id=gid,
                    missing_hours=demand - capacity,
                )
            )
        elif demand < capacity * 0.5:
            out.append(
                Conflict(
                    severity=Severity.INFO,
                    code="GROUP_UNDERUSED",
                    message=(
                        f"El grupo {group.label} sólo ocupa {demand} de {capacity} espacios; "
                        "quedarán muchas horas libres."
                    ),
                    group_id=gid,
                )
            )
    return out


def _check_eligibility(ctx: ProblemContext) -> List[Conflict]:
    """Toda materia con horas debe tener al menos un profesor que pueda impartirla."""
    out: List[Conflict] = []
    for gid in ctx.group_order:
        group = ctx.groups[gid]
        for c in group.curriculum:
            if c.weekly_hours == 0:
                continue
            elig = ctx.eligible.get((gid, c.subject_id), tuple())
            subject_name = ctx.subjects[c.subject_id].name
            if not elig:
                reason = (
                    "el grupo no tiene tutor asignado"
                    if c.assign_to_tutor
                    else "ningún profesor la tiene en su lista de materias"
                )
                out.append(
                    Conflict(
                        severity=Severity.ERROR,
                        code="NO_ELIGIBLE_TEACHER",
                        message=(
                            f"{subject_name} en {group.label}: {reason}. "
                            "Asigna la materia a algún profesor."
                        ),
                        group_id=gid,
                        subject_id=c.subject_id,
                        missing_hours=c.weekly_hours,
                    )
                )
                continue

            # Elegible pero sin un solo hueco compartido con el grupo.
            usable = set()
            for tid in elig:
                usable |= ctx.teacher_slots[tid] & ctx.group_slots[gid]
            if not usable:
                out.append(
                    Conflict(
                        severity=Severity.ERROR,
                        code="NO_COMMON_SLOT",
                        message=(
                            f"{subject_name} en {group.label}: la disponibilidad de sus profesores "
                            "no coincide con ningún horario hábil del grupo."
                        ),
                        group_id=gid,
                        subject_id=c.subject_id,
                        missing_hours=c.weekly_hours,
                    )
                )
            elif len(usable) < c.weekly_hours:
                out.append(
                    Conflict(
                        severity=Severity.ERROR,
                        code="NOT_ENOUGH_COMMON_SLOTS",
                        message=(
                            f"{subject_name} en {group.label} necesita {c.weekly_hours} h pero sólo "
                            f"existen {len(usable)} horarios compatibles con sus profesores."
                        ),
                        group_id=gid,
                        subject_id=c.subject_id,
                        missing_hours=c.weekly_hours - len(usable),
                    )
                )
    return out


def _check_max_per_day(ctx: ProblemContext) -> List[Conflict]:
    """`weekly_hours` no puede exceder `max_per_day` × días."""
    out: List[Conflict] = []
    n_days = len(ctx.request.grid.days)
    for gid in ctx.group_order:
        group = ctx.groups[gid]
        for c in group.curriculum:
            ceiling = c.max_per_day * n_days
            if c.weekly_hours > ceiling:
                out.append(
                    Conflict(
                        severity=Severity.ERROR,
                        code="MAX_PER_DAY_TOO_LOW",
                        message=(
                            f"{ctx.subjects[c.subject_id].name} en {group.label}: {c.weekly_hours} h "
                            f"semanales no caben con un tope de {c.max_per_day} h/día en {n_days} días."
                        ),
                        group_id=gid,
                        subject_id=c.subject_id,
                        missing_hours=c.weekly_hours - ceiling,
                    )
                )
    return out


def _subject_demand(ctx: ProblemContext) -> Dict[str, int]:
    demand: Dict[str, int] = defaultdict(int)
    for gid in ctx.group_order:
        for c in ctx.groups[gid].curriculum:
            demand[c.subject_id] += c.weekly_hours
    return demand


def _check_subject_capacity(ctx: ProblemContext) -> List[Conflict]:
    """
    Por materia: horas pedidas vs. techo de horas que pueden cubrir sus profesores.

    Es una cota superior optimista (ignora que un profesor se reparte entre materias),
    así que rebasarla es un error duro; quedarse justo debajo es una advertencia.
    """
    out: List[Conflict] = []
    demand = _subject_demand(ctx)
    for sid, hours in demand.items():
        if hours == 0:
            continue
        capacity = 0
        slot_capacity = 0
        for tid in ctx.teacher_order:
            if sid in ctx.teachers[tid].subject_ids:
                capacity += ctx.teacher_max_weekly[tid]
                slot_capacity += len(ctx.teacher_slots[tid])
        # Las materias impartidas por el tutor se saltan: su capacidad se valida
        # vía _check_eligibility, profesor por profesor.
        bound_to_tutor = all(
            c.assign_to_tutor
            for gid in ctx.group_order
            for c in ctx.groups[gid].curriculum
            if c.subject_id == sid and c.weekly_hours
        )
        if bound_to_tutor:
            continue
        effective = min(capacity, slot_capacity)
        if effective < hours:
            out.append(
                Conflict(
                    severity=Severity.ERROR,
                    code="SUBJECT_CAPACITY",
                    message=(
                        f"{ctx.subjects[sid].name}: la escuela requiere {hours} h semanales y sus "
                        f"profesores sólo pueden cubrir {effective} h "
                        f"(carga máxima {capacity} h, disponibilidad {slot_capacity} h). "
                        "Contrata apoyo o baja las horas del plan."
                    ),
                    subject_id=sid,
                    missing_hours=hours - effective,
                )
            )
        elif effective < hours * 1.1:
            out.append(
                Conflict(
                    severity=Severity.WARNING,
                    code="SUBJECT_CAPACITY_TIGHT",
                    message=(
                        f"{ctx.subjects[sid].name} va muy justa: {hours} h requeridas contra "
                        f"{effective} h disponibles. Es probable que falten horas por acomodar."
                    ),
                    subject_id=sid,
                )
            )
    return out


def _check_global_capacity(ctx: ProblemContext) -> List[Conflict]:
    total_demand = ctx.required_hours
    total_capacity = sum(ctx.teacher_max_weekly.values())
    if total_demand > total_capacity:
        return [
            Conflict(
                severity=Severity.ERROR,
                code="GLOBAL_CAPACITY",
                message=(
                    f"La escuela requiere {total_demand} horas-clase semanales y la plantilla "
                    f"suma {total_capacity} horas contratadas. Faltan {total_demand - total_capacity} h."
                ),
                missing_hours=total_demand - total_capacity,
            )
        ]
    if total_demand > total_capacity * 0.95:
        return [
            Conflict(
                severity=Severity.WARNING,
                code="GLOBAL_CAPACITY_TIGHT",
                message=(
                    f"Se usará más del 95 % de la plantilla ({total_demand}/{total_capacity} h). "
                    "Con tan poco margen es normal que el horario quede con huecos difíciles."
                ),
            )
        ]
    return []


def _check_teacher_availability(ctx: ProblemContext) -> List[Conflict]:
    out: List[Conflict] = []
    for tid in ctx.teacher_order:
        teacher = ctx.teachers[tid]
        available = len(ctx.teacher_slots[tid])
        if teacher.max_weekly_hours > 0 and available == 0:
            out.append(
                Conflict(
                    severity=Severity.WARNING,
                    code="TEACHER_NO_AVAILABILITY",
                    message=(
                        f"{teacher.name} tiene {teacher.max_weekly_hours} h contratadas pero su "
                        "matriz de disponibilidad no deja ningún horario libre."
                    ),
                    teacher_id=tid,
                )
            )
        elif available < teacher.max_weekly_hours:
            out.append(
                Conflict(
                    severity=Severity.INFO,
                    code="TEACHER_AVAILABILITY_BELOW_CONTRACT",
                    message=(
                        f"{teacher.name} sólo está disponible {available} h de las "
                        f"{teacher.max_weekly_hours} h contratadas."
                    ),
                    teacher_id=tid,
                )
            )
    return out
